applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.h

```diff
@@ -434,6 +451,77 @@ class AdjointFiniteDifferencingBaseElement : public Element
     ///@name Operations
     ///@{
 
+    template <typename TDataType>
+    void CalculateAdjointFieldOnIntegrationPoints(const Variable<TDataType>& rVariable, std::vector< TDataType >& rOutput, const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        KRATOS_WARNING_IF("CalculateAdjointFieldOnIntegrationPoints", OpenMPUtils::IsInParallel() != 0)
+                << "The call of this non omp-parallelized function within a parallel section should be avoided for efficiency reasons!" << std::endl;
+
+        const SizeType num_nodes = mpPrimalElement->GetGeometry().PointsNumber();
+        const SizeType dimension = mpPrimalElement->GetGeometry().WorkingSpaceDimension();
+        const SizeType num_dofs_per_node = (mHasRotationDofs) ?  2 * dimension : dimension;
+        const SizeType num_dofs = num_nodes * num_dofs_per_node;
+        Vector initial_state_variables;
+        initial_state_variables.resize(num_dofs);
+
+        Vector particular_solution = ZeroVector(num_dofs);
+        if(this->Has(ADJOINT_PARTICULAR_DISPLACEMENT))
+            particular_solution = this->GetValue(ADJOINT_PARTICULAR_DISPLACEMENT);
+
+        // Build vector of variables containing the DOF-variables of the primal problem
+        std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>*> primal_solution_variable_list;
+        primal_solution_variable_list.reserve(num_dofs_per_node);
+        primal_solution_variable_list.push_back(&DISPLACEMENT_X);
```

Comment:
i think using initilizerlist and tenary operator this could be written with much less lines:

```cpp
typedef std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>*> variable_pointer_vector;
const variable_pointer_vector primal_solution_variable_list = mHasRotationDofs ? 
    {&DISPLACEMENT_X, &DISPLACEMENT_Y, &DISPLACEMENT_Z, &ROTATION_X, &ROTATION_Y, &ROTATION_Z} 
    : {&DISPLACEMENT_X, &DISPLACEMENT_Y, &DISPLACEMENT_Z};
```

up to you if you like it.

Reply:
That looks nice :)